File: Prog/K_diff_pairs_in_array/K_diff_pairs.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

int findPairs(int* nums, int numsSize, int k) {
    int count = 0;
    
    // Create an array to store the frequency of each number
    int* numFreq = (int*)calloc(2001, sizeof(int));  // Assuming the range of numbers is [-1000, 1000]
    
    // Count the frequency of each number in the array
    for (int i = 0; i < numsSize; i++) {
        numFreq[nums[i] + 1000]++;
    }
    
    // Iterate through the array
    for (int i = 0; i < numsSize; i++) {
        if (k == 0) {
            if (numFreq[nums[i] + 1000] > 1) {
                count++;
                numFreq[nums[i] + 1000] = 0;  // Mark as visited to avoid duplicate pairs
            }
        } else {
            if (numFreq[nums[i] + k + 1000] > 0) {
                count++;
                numFreq[nums[i] + k + 1000] = 0;  // Mark as visited to avoid duplicate pairs
            }
        }
    }
    
    free(numFreq);  // Free the memory allocated for numFreq
    
    return count;
}

int main() {
    int nums[] = {3, 1, 4, 1, 5};
    int numsSize = sizeof(nums) / sizeof(nums[0]);
    int k = 2;
    int result = findPairs(nums, numsSize, k);
    printf("%d\n", result);
    
    return 0;
}

File: Prog/K_diff_pairs_in_array/K_diff_pairs.py
def findPairs(nums, k):
    num_freq = {}
    count = 0
    
    for num in nums:
        num_freq[num] = num_freq.get(num, 0) + 1
    
    for num in num_freq:
        if k == 0:
            if num_freq[num] > 1:
                count += 1
        else:
            if num + k in num_freq:
                count += 1
    
    return count

# Example usage
nums = [3, 1, 4, 1, 5]
k = 2
result = findPairs(nums, k)
print(result)
